datafusion/functions/src/regex/regexpcount.rs

```diff
@@ -0,0 +1,766 @@
+                    values.len(),
+                    flags_array.len()
+                )));
+            }
+
+            let mut regex_cache = HashMap::new();
+            Ok(Arc::new(Int64Array::from_iter_values(
+                values
+                    .iter()
+                    .zip(flags_array.iter())
+                    .map(|(value, flags)| {
+                        let pattern =
+                            compile_and_cache_regex(regex, flags, &mut regex_cache)?;
+                        count_matches(value, &pattern, start_scalar)
+                    })
+                    .collect::<Result<Vec<i64>, ArrowError>>()?,
+            )))
+        }
+        (true, false, true) => {
+            let regex = match regex_scalar {
+                None | Some("") => {
+                    return Ok(Arc::new(Int64Array::from(vec![0; values.len()])))
+                }
+                Some(regex) => regex,
+            };
+
+            let pattern = compile_regex(regex, flags_scalar)?;
+
+            let start_array = start_array.unwrap();
+
+            Ok(Arc::new(Int64Array::from_iter_values(
+                values
+                    .iter()
+                    .zip(start_array.iter())
+                    .map(|(value, start)| count_matches(value, &pattern, start))
+                    .collect::<Result<Vec<i64>, ArrowError>>()?,
+            )))
+        }
+        (true, false, false) => {
+            let regex = match regex_scalar {
+                None | Some("") => {
+                    return Ok(Arc::new(Int64Array::from(vec![0; values.len()])))
+                }
+                Some(regex) => regex,
+            };
+
+            let flags_array = flags_array.unwrap();
+            if values.len() != flags_array.len() {
+                return Err(ArrowError::ComputeError(format!(
+                    "flags_array must be the same length as values array; got {} and {}",
+                    values.len(),
+                    flags_array.len()
+                )));
+            }
+
+            let mut regex_cache = HashMap::new();
+            Ok(Arc::new(Int64Array::from_iter_values(
+                izip!(
+                    values.iter(),
+                    start_array.unwrap().iter(),
+                    flags_array.iter()
+                )
+                .map(|(value, start, flags)| {
+                    let pattern =
+                        compile_and_cache_regex(regex, flags, &mut regex_cache)?;
+
+                    count_matches(value, &pattern, start)
+                })
+                .collect::<Result<Vec<i64>, ArrowError>>()?,
+            )))
+        }
+        (false, true, true) => {
+            if values.len() != regex_array.len() {
+                return Err(ArrowError::ComputeError(format!(
+                    "regex_array must be the same length as values array; got {} and {}",
+                    values.len(),
```

Comment:
I would suggest aligning the parameters with the text by having regex_array.len() first